Tutorial update (#9940)
PrefectHQ/prefect: docs/tutorial/flows.md

---
description: Learn the basics of defining and running flows.
tags:
    - tutorial
    - getting started
    - basics
    - flows
    - logging
    - parameters
    - retries
---
## What is a flow?

A [flow](/concepts/flows/) is the basis of all Prefect workflows, it's a Python function decorated with a `@flow` decorator.

Some important points about flows:

> 1. All Prefect workflows are defined within the context of a flow.

Reviewer: Yeah, that stuck out to me as well. I think these points come from the old tutorial, so probably best to just exclude the term.

> - State transitions are recorded, allowing for flow execution to be observed.

Reviewer 2: ```suggestion
- It has a [state](/concepts/states/), which determines when it can run. Transitions between states are recorded, allowing for flow execution to be observed.
```

- Timeouts can be enforced to prevent unintentional, long-running workflows.
- Capture details about [flow runs](#flow-runs) such as run time, task tags, and final state.

## Run Your First flow:

> The simplest way get started with Prefect is to import and annotate your Python function with the [@flow](/api-ref/prefect/flows/#prefect.flows.flow) decorator. The script below fetches statistics about the main Prefect repository. Let's turn it into a Prefect flow:

Reviewer 2: ```suggestion
The simplest way get started with Prefect is to import and annotate your Python function with the `@flow` decorator. The script below fetches statistics about the main Prefect repository. Let's turn it into a Prefect flow:
```

```python hl_lines="2 5"
import httpx
from prefect import flow

@flow
def get_repo_info():
    url = "https://api.github.com/repos/PrefectHQ/prefect"
    response = httpx.get(url)
    response.raise_for_status()
    repo = response.json()
    print(f"PrefectHQ/prefect repository statistics 🤓:")
    print(f"Stars 🌠 : {repo['stargazers_count']}")
    print(f"Forks 🍴 : {repo['forks_count']}")

if __name__ == "__main__":
    get_repo_info()
```

> Running this flow in your terminal will result in some interesting output:

Reviewer 2: ```suggestion
Running this flow from your terminal will result in some interesting output:
```

<div class="terminal">
```bash
12:47:42.792 | INFO    | prefect.engine - Created flow run 'ludicrous-warthog' for flow 'get-repo-info'
PrefectHQ/prefect repository statistics 🤓:
Stars 🌠 : 12146
Forks 🍴 : 1245
12:47:45.008 | INFO    | Flow run 'ludicrous-warthog' - Finished in state Completed()
```
</div>

## Parameters

> As with any Python function, you can pass arguments. The positional and keyword arguments defined on your flow function are called [parameters](https://docs.prefect.io/2.10.15/concepts/flows/#parameters). Prefect will automatically perform type conversion by using any provided type hints. Let's make the repository a parameter:

Reviewer 2: ```suggestion
As with any Python function, you can pass arguments to a flow. The positional and keyword arguments defined on your flow function are called [parameters](https://docs.prefect.io/2.10.15/concepts/flows/#parameters). Prefect will automatically perform type conversion by using any provided type hints. Let's make the repository a parameter:
```

12:47:45.008 | INFO    | Flow run 'ludicrous-warthog' - Finished in state Completed()
```
</div>

Prefect can also capture `print` statements as info logs by specifying `log_prints=True` in your `flow` decorator (e.g. `@flow(log_prints=True)`).

## Retries

So far our script works, but in the future, the GitHub API may be temporarily unavailable or rate limited. [Retries](https://docs.prefect.io/2.10.15/concepts/flows/#flow-settings) help make our script more resilient. Let's add a retry functionality to our example above:
```python hl_lines="7"
import httpx
from prefect import flow, get_run_logger

@flow
def get_repo_info(
    repo_name: str = "PrefectHQ/prefect", retries=3, retry_delay_seconds=5
):
    url = f"https://api.github.com/repos/{repo_name}"
    response = httpx.get(url)
    response.raise_for_status()
    repo = response.json()
    logger = get_run_logger()
    logger.info(f"PrefectHQ/prefect repository statistics 🤓:")
    logger.info(f"Stars 🌠 : {repo['stargazers_count']}")
    logger.info(f"Forks 🍴 : {repo['forks_count']}")

if __name__ == "__main__":
    get_repo_info()
```

## [Next: Tasks](/tutorial/tasks/)

> As you have seen, adding a flow decorator converts our Python function to a resilient and observable workflow. In the next section you'll supercharge our flow by using tasks to break down the workflow's complexity and make it more performant.

Reviewer 2: ```suggestion
As you have seen, adding a flow decorator converts our Python function to a resilient and observable workflow. In the next section, you'll supercharge our flow by using tasks to break down the workflow's complexity and make it more performant.
```

---
description: Learn the basics of creating and running Prefect flows and tasks.
tags:
    - tutorial
    - getting started
    - basics
    - tasks
    - flows
    - subflows
---
## What is a flow?

A [flow](/concepts/flows/) is the basis of all Prefect workflows, it's a Python function decorated with a `@flow` decorator.

Some important points about flows:

1. All Prefect workflows are defined within the context of a flow.
2. Every Prefect workflow must contain at least one `flow` function that serves as the entrypoint for execution of the flow.

> 3. Flows can include calls to tasks as well as to child flows, which we call "subflows" in this context. At a high level, this is just like writing any other Python application: you organize specific, repetitive work into tasks, and call those tasks from flows.

Reviewer 2: ```suggestion
3. Flows can include calls to tasks as well as to child flows, which we call "subflows" in this context. Writing a Prefect workflow is just like writing any other Python application: you organize specific, repetitive work into tasks, and call those tasks from flows.
```

PrefectHQ/prefect repository statistics 🤓:
Stars 🌠 : 12146
Forks 🍴 : 1245
12:47:45.008 | INFO    | Flow run 'ludicrous-warthog' - Finished in state Completed()
```
</div>
## What can you do with flows?

### Retries

It helps provide for additional ways to respond on how your workflows fail, and offer more control on fail safe options for your workflow.
```python
import httpx
from prefect import flow

@flow # <--- This is a flow decorator!
def get_repo_info(retries = 3, retry_delay_seconds = 0.2):
    url = 'https://api.github.com/repos/PrefectHQ/prefect'
    api_response = httpx.get(url)
    if api_response.status_code == 200:
        repo_info = api_response.json()
        stars = repo_info['stargazers_count']
        forks = repo_info['forks_count']
        contributors_url = repo_info['contributors_url']
        print(f"PrefectHQ/prefect repository statistics 🤓:")
        print(f"Stars 🌠 : {stars}")
        print(f"Forks 🍴 : {forks}")
    else:
        raise Exception('Failed to fetch repository information.')

if __name__ == '__main__':
    get_repo_info()
```

The flow decorator lets you specify the number of retries.

### Next Steps

> To recap, simply adding an @flow decorator will convert a python function into an observed workflow. This pattern is coupled with a responsive user interface and fine tune orchestration features that are easy to add and quick to develop with.

Reviewer 2: ```suggestion
To recap, simply adding an `@flow` decorator will convert a python function into an observed workflow. This pattern is coupled with a responsive user interface and fine tune orchestration features that are easy to add and quick to develop with.
```

> On the next guide, we will showcase how to use tasks to in order to supercharge this github example even further and to help organize your complex workflows.

Reviewer 2: ```suggestion
In the next guide, we will showcase how to use tasks to in order to supercharge this github example even further and to help organize your complex workflows.
```
